ark/utils/metacluster_remap_gui/metaclusterdata.py

```diff
@@ -0,0 +1,128 @@
+        assert 1 in self.clusters.index and 0 not in self.clusters.index, "Cluster ids must be integer, starting with 1."  # noqa
+
+        self._output_mapping_filename = None
+        self._cached_metaclusters = None
+
+    @property
+    def output_mapping_filename(self):
+        return self._output_mapping_filename
+
+    @output_mapping_filename.setter
+    def output_mapping_filename(self, filepath):
+        self._output_mapping_filename = Path(filepath)
+
+    @property
+    def clusters_with_metaclusters(self):
+        df = self._clusters.join(self.mapping).sort_values(by='metacluster')
+        return df.iloc[:, self._marker_order + [max(self._marker_order)+1]]
+
+    @property
+    def clusters(self):
+        return self.clusters_with_metaclusters.drop(columns='metacluster')
+
+    @property
+    def metacluster_displaynames(self):
+        return [self.get_metacluster_displayname(mc) for mc in self.metaclusters.index]
+
+    @property
+    def metaclusters(self):
+        if self._cached_metaclusters is not None:
+            return self._cached_metaclusters
+        weighted_clusters = self.clusters.multiply(self.cluster_pixelcounts['count'], axis=0)
+        metacluster_pixelcounts = self.cluster_pixelcounts.join(self.mapping) \
+            .groupby('metacluster').aggregate('sum')
+        weighted_metaclusters = weighted_clusters.join(self.mapping) \
+            .groupby('metacluster').aggregate('sum') \
+            .divide(metacluster_pixelcounts['count'], axis=0)
+        self._cached_metaclusters = weighted_metaclusters
+        return weighted_metaclusters
+
+    @property
+    def linkage_matrix(self):
+        dist_matrix = cosine_similarity(self.clusters.T.values)
+        linkage_matrix = ward(dist_matrix)
+        return linkage_matrix
+
+    def get_metacluster_displayname(self, metacluster):
+        try:
+            return self._metacluster_displaynames_map[metacluster]
+        except KeyError:
+            return str(metacluster)
+
+    def cluster_in_metacluster(self, metacluster):
+        return list(self.mapping[self.mapping['metacluster'] == metacluster].index.values)
+
+    def which_metacluster(self, cluster):
+        return self.mapping.loc[cluster]['metacluster']
+
+    def new_metacluster(self):
+        return max(self.mapping['metacluster']) + 1
+
+    def remap(self, cluster, metacluster):
+        self.mapping.loc[cluster, 'metacluster'] = metacluster
+        self._cached_metaclusters = None
+
+    def change_displayname(self, metacluster, displayname):
+        self._metacluster_displaynames_map[metacluster] = displayname
+        self.save_output_mapping()
+
+    def save_output_mapping(self):
+        out_df = self.mapping.copy()
+        out_df['mc_name'] = [self.get_metacluster_displayname(mc) for mc in out_df['metacluster']]
+        out_df.to_csv(self.output_mapping_filename)
+
+    def set_marker_order(self, new_indexes):
+        assert set(new_indexes) == set(self._marker_order), \
```

Comment:
Same as above. We actually have a function that does this already: please take a look at `utils.misc_utils.verify_same_elements`.

Reply:
So (unlike the others, which i've got on the backlog to fix) this one really is an assert, not input validation. Meant to catch bugs at develop time when someone changes code. Could honestly be removed at this point if your style guide is to have no asserts. let me know.